lib/Settings/Settings.h

```diff
@@ -0,0 +1,81 @@
+/**
+  MQTT433gateway - MQTT 433.92 MHz radio gateway utilizing ESPiLight
+  Project home: https://github.com/puuu/MQTT433gateway/
+
+  The MIT License (MIT)
+
+  Copyright (c) 2017 Jan Losinski
+
+  Permission is hereby granted, free of charge, to any person
+  obtaining a copy of this software and associated documentation files
+  (the "Software"), to deal in the Software without restriction,
+  including without limitation the rights to use, copy, modify, merge,
+  publish, distribute, sublicense, and/or sell copies of the Software,
+  and to permit persons to whom the Software is furnished to do so,
+  subject to the following conditions:
+
+  The above copyright notice and this permission notice shall be
+  included in all copies or substantial portions of the Software.
+
+  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+  EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+  MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+  NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+  BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+  ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+  CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+  SOFTWARE.
+*/
+
+#ifndef MQTT433GATEWAY_SETTINGS_H
+#define MQTT433GATEWAY_SETTINGS_H
+
+#include <WString.h>
+#include <functional>
+
+enum SettingType { MQTT, RF_PROTOCOL, RF_ECHO, OTA };
+
+class SettingListener;
+
+class Settings {
+ public:
+  using SettingCallbackFn = std::function<void(const Settings &)>;
+
+  void onChange(SettingType setting, const SettingCallbackFn &callback);
+
+  String deviceName;
```

Comment:
With respect to the following commits and further changes (e.g. serialize), I think a list of many variables is not flexible enough. In python I would use a dict type. How about `std::map`?

Reply:
As they're queried that often I would not do that. Of course you could use the json object internally but this would cost performance. As we're on a limited device I would take the extra effort. On any platform I would go for a more generic container.

Reply:
Reading performance is a good argument. But how about caching critical setting in the responsible modules? We have the SettingHandler and do something similar with `logMode`/`rawMode`.

Reply:
We could do. But I think we'll not run into a situation with too many variables, as the feature-set of this firmware is pretty limited in the end. Also we need to know which callback to trigger on which change to the config. So we would need a some data structure to combine the setting with a type, etc. Also we need to use something like a variant type to store the type information of the data somewhere.

This should be easy to refactor later if we want, as the use of variables instead of generic containers gives us support from the compiler to find all the usages. For now I would leave it in this way.

Reply:
Ok, at the moment we can leave it as it is. But, we should still consider to refactor it later. The types, we can get from the default configuration (as long only json datatypes are interesting). Also possible could be to use nested objects, this can replace `std::bitset` and might be helpful when creating categoris in the web interface.